pkg/alias/aliaser.go

```diff
@@ -0,0 +1,86 @@
+package alias
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	cache "github.com/patrickmn/go-cache"
+	"github.com/sahilm/fuzzy"
+	"gorm.io/gorm"
+)
+
+type Alias struct {
+	NamedElementType string `gorm:"not null"`
+	NamedElementID   uint   `gorm:"not null"`
+	Alias            string
+}
+type Aliases []Alias
+
+func (a Aliases) String(i int) string {
+	return normalize(a[i].Alias)
+}
+func (a Aliases) Len() int {
+	return len(a)
+}
+
+type AliasLookupAgent struct {
+	db    *gorm.DB
+	cache *cache.Cache
+}
+
+func New(
+	db *gorm.DB,
+) AliasLookupAgent {
+	db = db.Set("gorm:auto_preload", true)
+
+	return AliasLookupAgent{
+		db: db,
+
+		// Refetch aliases from the database daily
+		cache: cache.New(24*time.Hour, 24*time.Hour),
+	}
+}
+
+func normalize(s string) string {
+	return strings.TrimSpace(strings.TrimSuffix(strings.ToUpper(s), "STATION"))
+}
+
+type AliasLookup interface {
+	FindNamedElementByRoughName(kind, name string) (*uint, error)
+}
+
+func (a AliasLookupAgent) getAliases(kind string) (Aliases, error) {
```

Comment:
This stuff is mostly copied from third-rail, but I tried to make it generic so that I wouldn't have to copy over all the models (line, direction, station). The three original implementations would first fetch the actual records and check for an exact match, then check the aliases. I couldn't find a good way to fetch the original records without copying a lot of code, so I made a change to third-rail to include the actual names among the aliases for all the records.